lib/pulse/lib/Pulse.Lib.Slice.fsti

```diff
@@ -0,0 +1,128 @@
+module Pulse.Lib.Slice
+open FStar.Tactics.V2
+open Pulse.Lib.Pervasives
+module SZ = FStar.SizeT
+module A = Pulse.Lib.Array
+
+val slice : Type0 -> Type0
+
+val len (#t: Type) : slice t -> SZ.t
+
+val pts_to (#t: Type) (s: slice t) (#[exact (`1.0R)] p: perm) (v: Seq.seq t) : slprop
+
+val pts_to_is_slprop2 (#a:Type) (x:slice a) (p:perm) (s:Seq.seq a)
+  : Lemma (is_slprop2 (pts_to x #p s))
+          [SMTPat (is_slprop2 (pts_to x #p s))]
+
+val pts_to_len (#t: Type) (s: slice t) (#p: perm) (#v: Seq.seq t) : stt_ghost unit emp_inames
+    (pts_to s #p v)
+    (fun _ -> pts_to s #p v ** pure (Seq.length v == SZ.v (len s)))
+
+val is_from_array (#t: Type) (a: array t) (p: perm) (s: slice t) : slprop
+
+val from_array (#t: Type) (mutb: bool) (a: array t) (#p: perm) (#v: Ghost.erased (Seq.seq t)) (alen: SZ.t {
+    SZ.v alen == A.length a /\
+    (mutb == true ==> p == 1.0R)
+}) : stt (slice t)
+    (A.pts_to a #p v)
+    (fun s ->
+        pts_to s #p v **
+        is_from_array a p s
+    )
+
+val to_array (#t: Type) (s: slice t) (#p: perm) (#v: Seq.seq t) (#a: array t) : stt_ghost unit emp_inames
+    (pts_to s #p v ** is_from_array a p s)
+    (fun _ -> A.pts_to a #p v)
+
+(* Written x.(i) *)
+val op_Array_Access
+        (#t: Type)
+        (a: slice t)
+        (i: SZ.t)
+        (#p: perm)
+        (#s: Ghost.erased (Seq.seq t){SZ.v i < Seq.length s})
+  : stt t
+        (requires
+            pts_to a #p s)
+        (ensures fun res ->
+            pts_to a #p s **
+            pure (res == Seq.index s (SZ.v i)))
+
+
+(* Written a.(i) <- v *)
+val op_Array_Assignment
+        (#t: Type)
+        (a: slice t)
+        (i: SZ.t)
+        (v: t)
+        (#s: Ghost.erased (Seq.seq t) {SZ.v i < Seq.length s})
+  : stt unit
+        (requires
+            pts_to a s)
+        (ensures fun res ->
+            pts_to a (Seq.upd s (SZ.v i) v))
+
+val share
+  (#a:Type)
+  (arr:slice a)
+  (#s:Ghost.erased (Seq.seq a))
+  (#p:perm)
+: stt_ghost unit emp_inames
+      (requires pts_to arr #p s)
+      (ensures fun _ -> pts_to arr #(p /. 2.0R) s ** pts_to arr #(p /. 2.0R) s)
+
+[@@allow_ambiguous]
+val gather
+  (#a:Type)
+  (arr:slice a)
+  (#s0 #s1:Ghost.erased (Seq.seq a))
+  (#p0 #p1:perm)
+: stt_ghost unit emp_inames
+      (requires pts_to arr #p0 s0 ** pts_to arr #p1 s1)
+      (ensures fun _ -> pts_to arr #(p0 +. p1) s0 ** pure (s0 == s1))
+
+val is_split (#t: Type) (s: slice t) (p: perm) (i: SZ.t) (left: slice t) (right: slice t) : slprop
+
+val is_split_is_slprop2 (#t: Type) (s: slice t) (p: perm) (i: SZ.t) (left: slice t) (right: slice t)
+  : Lemma (is_slprop2 (is_split s p i left right))
+          [SMTPat (is_slprop2 (is_split s p i left right))]
+
+noeq type slice_pair (t: Type) = | SlicePair: (fst: slice t) -> (snd: slice t) -> slice_pair t
```

Comment:
I added this type because Karamel refuses to extract functions returning F* pairs

Reply:
This is only an issue with the Rust backend, right?  In my experience, C-karamel does pairs just fine.

Reply:
if you could file a bug upstream with a precise description, that would be helpful, thank you